### PR TITLE
Safe RE_VA_ARG helpers

### PR DESCRIPTION
**Pull request by sreimers**
`RE_VA_ARG` helpers ensure type size safety by using the `_Generic` keyword (C11 only) to determine the size of the argument based on its type. This ensures that only arguments of the expected types are passed to the function, and prevents type mismatches that could lead to undefined behavior or security vulnerabilities.

### CODE REVIEW

File: test/test.c [hunk 1]
```diff
@@ -515,7 +515,7 @@ static int test_unit(const char *name, bool verbose)
 		for (i=0; i<RE_ARRAY_SIZE(tests); i++) {
 
 			if (verbose) {
-				re_printf("test %u -- %s\n",
+				re_printf("test %zu -- %s\n",
```
Comment (alfredh) [inline test/test.c:518]: some of the hunks in this patch are fixing the formatting strings.

it could be an idea to make a separate PR for these fixes ...

**Comment by alfredh** [comment]
this is great work!

Variable arguments is certainly a source for errors and security issues :)

**Comment by sreimers** [comment]
> I think this can be merged now..

Currently I have two open issues/todos:

- [ ] Fixing win32 x86 assert failure
- [ ] Create baresip PR to handle `src/config.c` (has many ARGS and uses `#ifdef` between them)

**Comment by alfredh** [comment]
this patch is quite large, around 900 lines.

since the patch is quite large and intrusive, would it be possible to split it up into smaller parts ?

**Comment by alfredh** [comment]
https://github.com/baresip/baresip/pull/2621

**Comment by sreimers** [comment]
Nice thanks! I will rebase this PR soon.

**Comment by alfredh** [comment]
https://github.com/baresip/baresip/pull/2627

File: include/re_types.h [hunk 2]
```diff
@@ -270,3 +270,76 @@ typedef size_t re_sock_t;
...
+#define RE_ARG_N3(N, ...) RE_ARG_##N(__VA_ARGS__)
+#define RE_ARG_N2(N, ...) RE_ARG_N3(N, __VA_ARGS__)
+#define RE_VA_ARGS(...) RE_ARG_N2(RE_ARG_VA_NUM(__VA_ARGS__), __VA_ARGS__)
+#endif
+
+#define RE_VA_ARG(ap, val, type, safe)                                        \
+	if ((safe)) {                                                         \
+		size_t sz = va_arg((ap), size_t);                             \
+		assert(sz && "RE_VA_ARG: no more arguments");                 \
+		assert(sz <= sizeof(type) &&                                  \
```
Comment (alfredh) [inline include/re_types.h:338]: Should we consider a warning instead of assert ?

Comment (sreimers) [reply include/re_types.h:338]: It would be nice, if a test fails if size is mismatched or no more args available.
cmake sets `-DNDEBUG`, on `CMAKE_BUILD_TYPE=Release`  builds and `assert()` is ignored.

File: src/fmt/time.c [hunk 3]
```diff
@@ -127,5 +127,5 @@ int fmt_timestamp(struct re_printf *pf, void *arg)
 #endif
 	(void)arg;
 
-	return re_hprintf(pf, "%02u:%02u:%02u.%03d", h, m, s, ms);
+	return re_hprintf(pf, "%02u:%02u:%02u.%03llu", h, m, s, ms);
```
Comment (alfredh) [inline src/fmt/time.c:130]: perhaps this can be a separate PR

File: src/mbuf/mbuf.c [hunk 4]
```diff
@@ -544,6 +544,27 @@ int mbuf_printf(struct mbuf *mb, const char *fmt, ...)
...
 
+/**
+ * Print a safe formatted string to a memory buffer
+ *
+ * @param mb  Memory buffer
+ * @param fmt Formatted string
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int _mbuf_printf_s(struct mbuf *mb, const char *fmt, ...)
```
Comment (alfredh) [inline src/mbuf/mbuf.c:555]: perhaps the mbuf changes can be added as a phase 2 ?

**Comment by alfredh** [comment]
The test "dtls_turn" is a combo test, I am not sure if its needed anymore.
All tests in retest are testing single modules, except "combo".


```
&& ./retest.exe -v -ri
regular tests:       Assertion failed: sz <= sizeof(unsigned long) && "RE_VA_ARG: arg is not compatible", file /home/runner/work/re/re/baresip-win32/re/src/fmt/print.c, line 349
Error: Signal (22) 0: test_dtls_turn (:0)
1: test_dtls_turn (:0)
2: raise (:0)
3: abort (:0)
4: assert (:0)
5: test_dtls_turn (:0)
6: test_dtls_turn (:0)
7: test_dtls_turn (:0)
8: test_dtls_turn (:0)
9: test_dtls_turn (:0)
```

Can you also log the printf string that is being printed ?

**Comment by alfredh** [comment]
I think you should try this patch on both 32-bits and 64-bits platforms.
For example size_t is 32-bits for 32-bits platforms.


Before the %zu fix in http, the test was failing on Mingw:


```
test 50 -- test_h265_packet
test 51 -- test_hash
test 52 -- test_hmac_sha1
test 53 -- test_hmac_sha256
test 54 -- test_http
test 55 -- test_http_loop
test 56 -- test_http_large_body
test 57 -- test_http_conn
... http_reqconn_send 639
Assertion failed: sz <= sizeof(unsigned long) && "RE_VA_ARG: arg is not compatible", file /home/alfredh/git/baresip-win32/re/s
rc/fmt/print.c, line 349

Error: Signal (22) 0:  (:0)
1:  (:0)
2: raise (:0)
3: _wassert (:0)
4: _assert (:0)
5: _assert (:0)
6: _assert (:0)
7: _assert (:0)
8: _assert (:0)
9: _assert (:0)

make: *** [Makefile:131: test] Error 3
```


perhaps you can rebase this PR to main HEAD ?

**Comment by sreimers** [comment]
Thanks @alfredh looks like its fixed now, will split up this PR.

**Comment by alfredh** [comment]
some more questions from me:

* Is it able to detect mismatch between signed and unsigned types (e.g. %u and int) ?
* Is there any runtime performance penalty when using the checks ?
* Is it an option to hide the feature behind a compile-time flag ?

**Comment by sreimers** [comment]
> * Is it an option to hide the feature behind a compile-time flag ?

A application can always opt-out with "-DDISABLE_RE_ARG=1" or while including headers, the interface is flexible. So its possible to opt-out within specific parts of a application with `#define DISABLE_RE_ARG 1` before loading `re.h` if necessary.

> * Is there any runtime performance penalty when using the checks ?

Yes but these are very marginal (I can provide some google benchmarks). In release the assert checks are not used and incompatible or to many args simply ignored.

> * Is it able to detect mismatch between signed and unsigned types (e.g. %u and int) ?

I will have a look on this. not sure.

File: include/re_types.h (hunk 2 above)
Comment (alfredh) [reply include/re_types.h:338]: Another approach is to return an error code if the format is wrong.

This means that e.g. mbuf_printf would return the error code.

**Comment by alfredh** [comment]
could you please rebase on top of HEAD again ?

I did some testing of this patch, with baresip and it did not print any warnings...

File: include/re_types.h (hunk 2 above)
Comment (sreimers) [reply include/re_types.h:338]: > Another approach is to return an error code if the format is wrong.

Most printf code ignores the return code I think, but its already set and exited here:
```c
		if (!sz || sz > sizeof(type)) {                               \
			err = EINVAL;                                         \
			goto out;                                             \
		}               
```
With Release build type assert is not called (I can make this more visible with `#ifndef RELEASE`)

File: src/mbuf/mbuf.c (hunk 4 above)
Comment (sreimers) [reply src/mbuf/mbuf.c:555]: Ok will exclude this commit.

**Comment by alfredh** [comment]
it is looking better now. Should we merge it ?

**Comment by juha-h** [comment]
Looks like after this commit, baresip-studio project Android build stopped working:

**Comment by juha-h** [comment]
The build worked when I commented out this:
```
//#if defined(__ARM_ARCH_7A__)
//    #define long long long
//#endif
```
But I don't have any armeabi-v7a devices to test if the app actually works.

**Comment by sreimers** [comment]
> #define long long long

Where is this defined?

**Comment by juha-h** [comment]
Sebastian Reimers writes:

> > #define long long long
> 
> Where is this defined?

It was defined in my app and was needed before this commit.  Otherwise
32 bit build didn't work.
